File: code/ex402.java
// Discard the previous digit if it is larger than the current digit

import java.util.*;

class Solution {
    public String removeKdigits(String num, int k) {
        Stack<Character> stack = new Stack<>();
        
        for (int i = 0; i < num.length(); i++) {
            while (!stack.isEmpty() && stack.peek() > num.charAt(i) && k-- > 0) {
                stack.pop();
            }
            stack.push(num.charAt(i));
        }
        
        // If k still > 0, delete rest digits from the back since the last few digits are larger in the stack
        while (k-- > 0) {
            stack.pop();
        }
        
        // System.out.println(stack);
        
        // Convert the stack to string
        StringBuilder sb = new StringBuilder();
        
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }
        sb.reverse();
        
        // Delete 0 from the head of the stringbuilder
        while (sb.length() > 1 && sb.charAt(0) == '0')
            sb.deleteCharAt(0);
        
        return (sb.length() == 0) ? "0" : sb.toString();
    }
}

File: code/ex402.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        int n = num.length();
        
        stack<char> stk;
        for (char c : num) {
            while (!stk.empty() && stk.top() > c && k > 0) {
                stk.pop();
                k--;
            }
            stk.push(c);
        }
        
        // Remove rest of the k chars from the end
        while (k > 0) {
            stk.pop();
            k--;
        }
        
        // Get result
        string res = "";
        while (!stk.empty()) {
            res = stk.top() + res;
            stk.pop();
        }
        
        // Remove leading zeros
        int start = 0;
        while (start < res.length() && res[start] == '0')    start++;
        
        // Get final answer
        string ret = res.substr(start);
        return ret == "" ? "0" : ret;
    }
};